### PR TITLE
Fix equality for Arr<A>

### LINKED ISSUE #286: Stackoverflow in Arr<Arr<double>>.Equals
image

The test case that generates the stack overflow is

        [Fact]
        public void ArrShouldNotStackOverflowOnEquals()
        {
            Arr<Arr<double>> arr = default;

            arr.Equals( arr ).Should().BeTrue();
        }

Wierd!!

### PR DESCRIPTION
**Pull request by bradphelan**
Fixes #286 

There probably should be more testing on equality for all the types. This just fixes ``Arr<T>``

### CODE REVIEW

File: LanguageExt.Core/DataTypes/Arr/Arr.cs [hunk 1]
```diff
@@ -449,6 +449,22 @@ public Arr<B> Map<B>(Func<A, B> map)
         public override bool Equals(object obj) =>
             obj is Arr<A> && Equals((Arr<A>)obj);
 
+        [Pure]
+        public bool Equals(Arr<A> obj) =>
+            SafeEquals(Value, obj.Value);
```
Comment (louthy) [inline LanguageExt.Core/DataTypes/Arr/Arr.cs]: Should probably just call `default(MArr<A>).Equals(this, obj);`

**Comment by bradphelan** [comment]
doesnt that allocate and then deallocate a backing array every time you do
that?

**Comment by louthy** [comment]
@bradphelan 

> doesnt that allocate and then deallocate a backing array every time you do
that?

Nope. `MArr<A>` is the 'class instance' for the `Arr<A>` data-type, it doesn't allocate anything, just calls: `Enumerable.SequenceEqual(x, y)`

**Comment by louthy** [comment]
@bradphelan Added my own fixes for this, so closing.  Thanks for the heads up (Y).

**Comment by bradphelan** [comment]
no i mean that Arr<T> is a struct but is backed by an array reference which
needs to allocate on heap. Therefore default(Arr<T>) will allocate an array
every time it is called.

**Comment by louthy** [comment]
It's default(MArr) not default(Arr) - the Monad/Eq/Ord type-class instance for Arr.

**Comment by bradphelan** [comment]
my bad! thanks for the clarification

**Pull request closed without merging**